chore(perps): Refactor HL subscription service

michalconsensys opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->

## **Description**

This PR refactors the HyperLiquid subscription service to drop subscription fallbacks.

This refactor was raised during the offsite, as it is needed as a result of the hotfix we had to deploy to address Perps loading issue during the weekend. The initial PR https://github.com/MetaMask/metamask-mobile/pull/23753 introduced a fallback mechanism that would create new subscriptions when the properties in the data coming from the old subscriptions were missing. Given that these properties will be always missing, this PR uses the new subscriptions as a default.



## **Changelog**

<!--
If this PR is not End-User-Facing and should not show up in the CHANGELOG, you can choose to either:
1. Write `CHANGELOG entry: null`
2. Label with `no-changelog`

If this PR is End-User-Facing, please write a short User-Facing description in the past tense like:
`CHANGELOG entry: Added a new tab for users to see their NFTs`
`CHANGELOG entry: Fixed a bug that was causing some NFTs to flicker`

(This helps the Release Engineer do their job more quickly and accurately)
-->

CHANGELOG entry: null

## **Related issues**

Fixes:

## **Manual testing steps**

```gherkin
Feature: my feature name

  Scenario: user [verb for user action]
    Given [describe expected initial app state]

    When user [verb for user action]
    Then [describe expected outcome]
```

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

No visible change

### **After**

No visible change

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.


<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Replaces HIP‑3 webData3 fallback with per‑DEX clearinghouseState/openOrders subscriptions, adds `triggerPrice` to orders, and hardens TP/SL extraction with price-based fallback and cache re-use; tests updated accordingly.
> 
> - **Perps Subscription Service (HIP‑3)**:
>   - Switch to individual subscriptions: `clearinghouseState` (positions/account) and `openOrders` (orders); use `webData3` only for OI caps.
>   - Remove fallback caches/handlers; centralize aggregation and cache updates; update cleanup/restore logic to handle new subs.
>   - Re-extract TP/SL from cached orders on `clearinghouseState` updates to prevent flicker.
>   - Update hashing/change-detection and notify paths; broaden reconnection restoration across subtypes.
> - **TP/SL Processing**:
>   - New shared helper to extract TP/SL from orders; adds price-based fallback for ambiguous `Trigger` types (long/short aware) and counts TP/SL.
> - **Order Model & Adapter**:
>   - Extend `Order` with `triggerPrice`.
>   - `adaptOrderFromSDK` sets `triggerPrice` (uses `triggerPx`) and prioritizes `limitPx` for price; child TP/SL preserved.
> - **Tests**:
>   - Overhaul to reflect new HIP‑3 flow: assert `clearinghouseState`/`openOrders` usage; OI caps via `webData3`.
>   - Replace timeouts with `jest.runAllTimersAsync()`; add coverage for TP/SL classification, cache re-use, short/long logic, L2 book handling, and feature-flag updates.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit cd9bbb06fd366152dc869c3e021f4cd17a6a8a72. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

michalconsensys committed:
wip

michalconsensys committed:
chore: remove comments

michalconsensys committed:
chore: fix tests

michalconsensys committed:
chore: improve coverage

michalconsensys committed:
chore: fix ts

michalconsensys committed:
chore: CR

michalconsensys committed:
Merge branch 'main' into chore/perps/subscriptions-refactor

michalconsensys committed:
Merge branch 'main' into chore/perps/subscriptions-refactor

abretonc7s committed:
Merge branch 'main' into chore/perps/subscriptions-refactor

abretonc7s commented on app/components/UI/Perps/services/HyperLiquidSubscriptionService.ts:
```diff
@@ -1172,25 +1060,80 @@ export class HyperLiquidSubscriptionService {
+              >();
+
+              cachedOrders.forEach((order) => {
+                // Check if this is a trigger order (TP/SL) by checking isTrigger flag
```
I think this logic could be extracted and is similar to extractTPSLFromOrders

abretonc7s reviewed (changes requested):
Optional but  there must be a way to avoid having the setTimeout in the unit tests which may make them fragile

michalconsensys committed:
refactor: unify TP/SL extraction logic in extractTPSLFromOrders method

michalconsensys committed:
test(perps): replace setTimeout waits with jest fake timers in HyperLiquidSubscriptionService tests

- Add jest.useFakeTimers() in beforeEach and jest.useRealTimers() in afterEach
- Replace 77 occurrences of setTimeout-based waits with jest.runAllTimersAsync()
- Fix race condition test that had internal setTimeout causing deadlock
- Makes tests deterministic and not fragile due to timing issues

michalconsensys committed:
fix: update cache

michalconsensys committed:
chore: fix test

michalconsensys committed:
fix(perps): add price-based fallback for TP/SL re-extraction from cached orders

The extractTPSLFromOrders method's cached orders branch was missing the
price-based fallback logic for ambiguous 'Trigger' order types. When orders
had detailedOrderType of 'Trigger' (without 'Take Profit' or 'Stop'), neither
classification matched, causing TP/SL data to disappear from positions when
clearinghouseState updates triggered re-extraction from cached orders.

Added the same price-based fallback that exists in the raw SDK branch:
- For long positions: trigger price > entry price = Take Profit
- For short positions: trigger price < entry price = Take Profit
- Otherwise classified as Stop Loss

Also added test to verify TP/SL persistence across clearinghouseState updates.

michalconsensys commented on app/components/UI/Perps/services/HyperLiquidSubscriptionService.ts:
```diff
@@ -442,6 +435,48 @@ export class HyperLiquidSubscriptionService {
+      });
+
+      return { tpslMap, tpslCountMap, processedOrders: cachedProcessedOrders };
+    }
```
Resolved in https://github.com/MetaMask/metamask-mobile/pull/24015/commits/130f8c3576730713a37ac541116ab78cff1fcbf2

abretonc7s committed:
Merge branch 'main' into chore/perps/subscriptions-refactor

michalconsensys committed:
fix(perps): add price-based fallback for TP/SL count calculation in raw SDK orders

The raw SDK order processing path was missing price-based fallback
classification for ambiguous 'Trigger' type orders. This caused
inconsistent TP/SL counts between the openOrders callback (count=0)
and subsequent clearinghouseState updates using cached orders (count=1).

Now both paths use the same classification logic:
1. First check explicit order type (Take Profit/Stop)
2. If ambiguous, use price-based detection based on position direction
3. Use classified values for both counts and price assignment

michalconsensys committed:
Merge branch 'main' into chore/perps/subscriptions-refactor

michalconsensys committed:
Merge branch 'main' into chore/perps/subscriptions-refactor
